src/subscription/RelaySubscriptionObservable.js

```diff
@@ -0,0 +1,209 @@
+  RelayMutationConfig,
+  SubscriptionCallbacks,
+  Subscription,
+  SubscriptionResult,
+} from 'RelayTypes';
+
+const base62 = require('base62');
+const invariant = require('invariant');
+const RelayConnectionInterface = require('RelayConnectionInterface');
+const RelayQuery = require('RelayQuery')
+const RelaySubscriptionQuery = require('RelaySubscriptionQuery');
+const RelaySubscriptionRequest = require('RelaySubscriptionRequest');
+
+const {CLIENT_SUBSCRIPTION_ID} = RelayConnectionInterface;
+let subscriptionIDCounter = 0;
+
+/**
+ * An Rx Observable representing the results of a RelaySubscription.
+ * Subscribers are notified as follows:
+ *
+ * `onNext`: Called with the latest results of a RelaySubscription. Results will be of type SubscriptionResult.
+ * `Relay.Store.NetworkLayer.sendSubscription`
+ * - Called synchronously on first `subscribe()`.
+ * - Called whenever the results of the RelaySubscription are received.
+ *
+ * `onError`: Called if the RelaySubscription gets an error and no more results will be received.
+ *
+ * `onCompleted`: Called when the RelaySubscription has completed successfully and no more results will be recieved.
+ *
+ * @see http://reactivex.io/documentation/observable.html
+ */
+class RelaySubscriptionObservable {
+  _observer: RelaySubscriptionObserver;
+  _storeData: RelayStoreData;
+  _subscription: RelaySubscription;
+  _subscriptionCallbacks: Array<SubscriptionCallbacks<SubscriptionResult>>;
+  _subscriptionConfigs: ?Array<RelayMutationConfig>;
+  _subscriptionCount: number;
+  _subscriptionDisposable: ?Subscription;
+  _subscriptionQuery: ?RelayQuery.Subscription;
+
+  constructor(
+    observer: RelaySubscriptionObserver,
+    storeData: RelayStoreData,
+    subscription: RelaySubscription,
+  ) {
+    this._observer = observer;
+    this._storeData = storeData;
+    this._subscription = subscription;
+    this._subscriptionCallbacks = [];
+    this._subscriptionConfigs = null;
+    this._subscriptionCount = 0;
+    this._subscriptionDisposable = null;
+    this._subscriptionQuery = null;
+  }
+
+  /**
+   * Subscribe to the subscription
+   */
+  subscribe(callbacks: SubscriptionCallbacks<SubscriptionResult>): Subscription {
+    this._subscriptionCount++;
+    const subscriptionIndex = this._subscriptionCallbacks.push(callbacks);
+    const subscribeDisposable = {
+      dispose: () => {
+        invariant(
+          this._subscriptionCallbacks[subscriptionIndex],
+          'RelayQueryResultObservable: Subscriptions may only be disposed once.',
+        );
+        delete this._subscriptionCallbacks[subscriptionIndex];
+        this._subscriptionCount--;
+        if (this._subscriptionCount === 0) {
+          this._observer.unobserve(this._subscription);
+        }
+      }
+    }
+
+    if (this._subscriptionCount === 1) {
+      this._observe();
+    }
+
+    return subscribeDisposable;
+  }
+
+  /**
+   * Dispose underlying RelaySubscription subscription
+   */
+  unobserve(): void {
```

Comment:
That line is for a private method, used to handle the case when a subscriber is disposed. Is this intended to be a public method that a developer would call? Or a public method that the subscription framework would call?

Aside: this PR is really big, it would be a lot easier to review as a series of incremental changes.